# Data Cleaning with Pandas

In this notebook we'll go through a few basic data cleaning steps that should be performed on all new datasets where necessary.

We'll go through the process with both the `orders` and `orderlines` datasets. You can then practice these skills by cleaning the `products` dataset yourself

In [1]:
import pandas as pd

In [2]:
# orders.csv
path = "C:/Users/ruben/WBS/Data_Science/Section_3/Data_Cleaning_Storytelling_with_Pandas_and_Matplotlib/data/raw_data/orders.csv"
orders = pd.read_csv(path)

# orderlines.csv
path = "C:/Users/ruben/WBS/Data_Science/Section_3/Data_Cleaning_Storytelling_with_Pandas_and_Matplotlib/data/raw_data/orderlines.csv"
orderlines = pd.read_csv(path)

One of the best ways to begin data cleaning is by exploring using `.info()`. This will tell us:
* The shape of the DataFrame
* The names of the columns
* If there are any missing values
* The datatypes of the columns

By exploring the missing values and correcting any incorrect datatypes, we often come across inconsistencies in our data.

Beyond this, we should also have a **check for any duplicate rows**. 

Let's first deal with the duplicates, as it's nice and easy, then we'll explore what `.info()` has to tell us.

## 1.&nbsp; Duplicates
We can check for duplicates using the pandas [.duplicated()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) method. 

We can then delete these rows, if we wish, using [.drop_duplicates()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html)

In [4]:
# orders
orders.duplicated().sum()

0

In [5]:
# orderlines
orderlines.duplicated().sum()

0

We have no duplicate rows in either DataFrame. Easy, there is no problem to solve. Normally though, if there were some duplicates, we'd drop the extra rows.

# 2.&nbsp; `.info()`

In [6]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


* `total_paid` has 5 missing values
* `created_date` should become datetime datatype

In [7]:
orderlines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


* `date` should be a datetime datatype
* `unit_price` should be a float datatype

## 3.&nbsp; Missing values

### 3.1.&nbsp; Orders
* `total_paid` has 5 missing values

In [8]:
print(f"5 missing values represents {((orders.total_paid.isna().sum() / orders.shape[0])*100).round(5)}% of the rows in our DataFrame")

5 missing values represents 0.0022% of the rows in our DataFrame


As there is such a tiny amount of missing values, we will simply delete these rows, as we have enough data without them.

In [9]:
orders = orders.loc[~orders.total_paid.isna(), :]

Should you have a significant number of missing values in the future, you have a choice: 
+ you can impute the values
+ you can take the values from other DataFrames, if they are present there
+ you can delete the values
+ or any number of other creative solutions

Please, always consider how much time you have on your project, and what impact your method of choice will have on your final assesment.

### 3.2.&nbsp; Orderlines
There are no missing values in `orderlines`

## 4.&nbsp; Datatypes

### 4.1.&nbsp; Orders
* `created_date` should become datetime datatype

In [10]:
orders["created_date"] = pd.to_datetime(orders["created_date"])

### 4.1.&nbsp; Orderlines
* `date` should be a datetime datatype
* `unit_price` should be a float datatype

#### 4.1.1.&nbsp; `date`

In [11]:
orderlines["date"] = pd.to_datetime(orderlines["date"])

#### 4.1.2.&nbsp;`unit_price`

In [12]:
orderlines["unit_price"] = pd.to_numeric(orderlines["unit_price"])

ValueError: Unable to parse string "1.137.99" at position 6

As you can see when we try to convert `unit_price` to a numerical datatype, we receive a `ValueError` telling us that pandas doesn't understand the number `1.137.99`. This is probably because numbers cannot have 2 decimal points. Let's see if there are any other numbers like this.

In [14]:
orderlines.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()

False    257814
True      36169
Name: unit_price, dtype: int64

Looks like over 36000 rows in `orderlines` are affected by this problem. Let's work out how much that is as a percentage of our total data.

In [15]:
two_dot_percentage = ((orderlines.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()[1] / orderlines.shape[0])*100).round(2)
print(f"The 2 dot problem represents {two_dot_percentage}% of the rows in our DataFrame")

The 2 dot problem represents 12.3% of the rows in our DataFrame


This is a bit of a tricky decision as 12.3% is a significant amount of our data... and we might even end up losing a larger portion of our data than this too. For the moment we will delete the rows as we only have 2 weeks for this project and I'd like some quick, accurate results to show. If we have time at the end, we can come back and investigate this problem further, maybe there's a solution?

Each row of `orderlines` represents a product in an order. For example, if order number 175 contained 3 seperate products, then order 175 would have 3 rows in `orderlines`, one row for each of the products. If 2 of those products have 'normal' prices (14.99, 15.85) and 1 has a price with 2 decimal points (1.137.99), we need to remove the whole order and not just the affected row. If we only remove the row with 2 decimal places then any later analysis about products and prices could be misleading.

We therefore need to find the order numbers associated with the rows that have 2 decimal points, and then remove all the associated rows.

In [16]:
two_dot_order_ids_list = orderlines.loc[orderlines.unit_price.str.contains("\d+\.\d+\.\d+"), "id_order"]
orderlines = orderlines.loc[~orderlines.id_order.isin(two_dot_order_ids_list)]

In [17]:
orderlines.shape[0]

216250

We still have 216250 rows in orderlines to work with. This should be more than enough for our evaluation.

Now that all of the 2 decimal point prices have been removed, let's try again to convert the column `unit_price` to the correct datatype.

In [19]:
orderlines["unit_price"] = pd.to_numeric(orderlines["unit_price"])

It worked perfectly

# Challenge: Clean the `products` DataFrame
Now it's your turn. Use the lessons you learnt above and clean the products DataFrame. You don't have to copy exactly what we did. Think about the consequences of your actions, sometimes it is ok to delete rows, other times you may wish to come up with more creative solutions.

In [3]:
# products.csv
path = "C:/Users/ruben/WBS/Data_Science/Section_3/Data_Cleaning_Storytelling_with_Pandas_and_Matplotlib/data/raw_data/products.csv"
products = pd.read_csv(path)

In [20]:
# your code here
products.head()

sku                                           name  \
0  RAI0007              Silver Rain Design mStand Support   
1  APP0023              Apple Mac Keyboard Keypad Spanish   
2  APP0025               Mighty Mouse Apple Mouse for Mac   
3  APP0072  Apple Dock to USB Cable iPhone and iPod white   
4  KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   

                                                desc  price promo_price  \
0       Aluminum support compatible with all MacBook  59.99     499.899   
1          USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                             mouse Apple USB cable.     59     569.898   
3              IPhone dock and USB Cable Apple iPod.     25     229.997   
4  2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   

   in_stock      type  
0         1      8696  
1         0  13855401  
2         0      1387  
3         0      1230  
4         1      1364

In [22]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


### Look for Duplicates

In [21]:
# your code here
products.duplicated().sum()

8746

In [23]:
products.drop_duplicates(inplace=True)

In [24]:
products.duplicated().sum()

0

### Look for Missing values


In [27]:
# your code here
products.desc.isna().sum()

7

In [28]:
products.price.isna().sum()

46

In [29]:
products.type.isna().sum()

50

* `desc` has 7 missing values
* `price` has 46 missing values
* `type` has 50 missing values

#### `desc`

7 is a very small number to have missing, let's have a closer look

In [30]:
products.loc[products.desc.isna(), :]

sku                                               name desc  \
16126  WDT0211-A  Open - Purple 2TB WD 35 "PC Security Mac hard ...  NaN   
16128  APP1622-A  Open - Apple Smart Keyboard Pro Keyboard Folio...  NaN   
17843    PAC2334             Synology DS718 + NAS Server | 10GB RAM  NaN   
18152  KAN0034-A  Open - Kanex USB-C Gigabit Ethernet Adapter Ma...  NaN   
18490    HTE0025  Hyper Pearl 1600mAh battery Mini USB Mirror an...  NaN   
18612    OTT0200    OtterBox External Battery Power Pack 20000 mAHr  NaN   
18690  HOW0001-A  Open - Honeywell thermostat Lyric zonificador ...  NaN   

           price promo_price  in_stock      type  
16126        107     814.659         0      1298  
16128  1.568.206   1.568.206         0      1298  
17843     566.35   5.659.896         0  12175397  
18152      29.99     237.925         0      1298  
18490      24.99       22.99         1      1515  
18612      79.99       56.99         1      1515  
18690     199.99   1.441.174         0  11905404

We have 2 choices here:
* We can quickly and easily remove these rows. 
* Or, alternatively, the products names here are quite descriptive, so I'm tempted to just copy them to the description column, so that there is a description if we later want utilise this column. I wouldn't recommend this if this DataFrame was the source of truth for our website. But this is not the case here, and I'd we're not faking any information (guessing a price or so), so I'm happy with this option

In [31]:
products.loc[products.desc.isna(), "desc"] = products.loc[products.desc.isna(), "name"]

Did you also notice above that we have the dreaded two decimal point problem in both the `price` and `promo_price` columns? We can also see prices with 3 decimal places, prices should have 2 decimal places: this gives us more cause for concern

#### `price`

In [32]:
products.price.isna().sum()

46

In [33]:
print(f"The missing values in price are {((products.price.isna().sum() / products.shape[0]) * 100).round(2)}% of all rows in the DataFrame")

The missing values in price are 0.43% of all rows in the DataFrame


Let's simply delete these rows to ensure that we can trust the numbers in our final DataFrame. Afterall, the price is very important when investigating discounts.

In [34]:
products = products.loc[~products.price.isna(), :]

In [35]:
products.price.isna().sum()

0

### Check / Change Data types

We saw from looking at the output of `.info()` that both `price` and `promo_price` have been stored as objects and not as a numerical datatypes. We also saw while solving other problems that both columns have some prices with 3 decimal places and others with 2 decimal points - the latter will prevent us from converting the datatype to numerical, so first we must investigate and solve these problems.

#### `price`

First, let's see how many values are affected by the 2 problems mentioned above.

In [38]:
# your code here
price_problems_number = products.loc[(products.price.astype(str).str.contains("\d+\.\d+\.\d+")) | (products.price.astype(str).str.contains("\d+\.\d{3,}")), :].shape[0]
price_problems_number

542

In [39]:
print(f"The column price has in total {price_problems_number} wrong values. This is {round(((price_problems_number / products.shape[0]) * 100), 2)}% of the rows of the DataFrame")

The column price has in total 542 wrong values. This is 5.15% of the rows of the DataFrame


5.15% is a reasonable amount of our data. However, the price column will be important to understanding discounts, so I'd like it to be very trustworthy as we are basing business decisions on it. Therefore, we'll delete these rows

In [40]:
products = products.loc[(~products.price.astype(str).str.contains("\d+\.\d+\.\d+"))&(~products.price.astype(str).str.contains("\d+\.\d{3,}")), :]

To complete our task, let's convert the column to a numeric datatype

In [41]:
products["price"] = pd.to_numeric(products["price"])

In [46]:
price_problems_number = products.loc[(products.price.astype(str).str.contains("\d+\.\d+\.\d+")) | (products.price.astype(str).str.contains("\d+\.\d{3,}")), :].shape[0]

In [47]:
print(f"The column price has in total {price_problems_number} wrong values. This is {round(((price_problems_number / products.shape[0]) * 100), 2)}% of the rows of the DataFrame")

The column price has in total 0 wrong values. This is 0.0% of the rows of the DataFrame


#### `promo_price`

Again, let's begin by seeing how many values are affected by the 2 problems

In [48]:
promo_problems_number = products.loc[(products.promo_price.astype(str).str.contains("\d+\.\d+\.\d+"))|(products.promo_price.astype(str).str.contains("\d+\.\d{3,}")), :].shape[0]
promo_problems_number

9232

In [49]:
print(f"The column promo_price has in total {promo_problems_number} wrong values. This is {round(((promo_problems_number / products.shape[0]) * 100), 2)}% of the rows of the DataFrame")

The column promo_price has in total 9232 wrong values. This is 92.39% of the rows of the DataFrame


WOW!!! That's a lot of wrong data. Let's have a quick investigate to check that's correct. We'll make a DataFrame by copy-pasting the code we used above and then look at a large sample to check that all the numbers in the `promo_price` column really have either 2 decimal points or 3 decimal places.

In [50]:
promo_price_df = products.loc[(products.promo_price.astype(str).str.contains("\d+\.\d+\.\d+"))|(products.promo_price.astype(str).str.contains("\d+\.\d{3,}")), :]
promo_price_df.sample(50)

sku                                               name  \
17729    NIE0010                 Segway personal transport MiniLite   
17839    PAC2330  Synology DS718 + NAS Server | 6GB RAM | 16TB (...   
11928  NTE0025-A   Open - NewerTech Mini DisplayPort to DVI Adapter   
13579    APP1712  Apple Watch Series 2 38mm Stainless Steel Case...   
671      SAN0046                SanDisk Ultra 64GB USB 3.0 PenDrive   
10692    PAC1778            Synology Pack DS216SE | Seagate 8TB NAS   
16078    AP20186          Like new - Apple iPhone 32GB Rose Gold 6s   
1522     PAC1823  Synology DS1515 + l l 8GB RAM 50TB (5x10TB) Se...   
14506  SYN0132-A       Open - Synology RT1900AC Wifi Router AC 1900   
18895    APP2366  Apple Macbook Pro 13 "Core i7 25GHz | 16GB | 1...   
13020    CRU0053                         Disk 1TB SSD Crucial MX300   
16706    APP2438  Apple Macbook 12 "Core i7 14GHz | 8GB | 512GB ...   
19076    TWS0126          Twelve South BookBook iPhone Case Brown X   
10822    ICA0075  Incase Sleeve Tensaerlite ICON 13 Macbook Air ...   
14859    TAM0012         Tangram Comba Smart Led Rope Chrome Size M   
15460    REP0357                     Speaker repair inside iPad Air   
645      SEA0057  Seagate 500GB SSHD Hybrid disc 25 "Mac and PC ...   
18699  NTE0111-A  Open - NewerTech NuGuard Snap-on Case MacBook ...   
1530     PAC1817            Synology DS1515 + l 40TB (5x8TB) WD Red   
18754  GTE0096-A  Open - G-Technology G-DRIVE HDD 10TB 7200rpm T...   
3016     APP1166                        Apple iPhone 6S 16GB Silver   
2299     TRA0027  Transcend JetDrive expansion SSD 725 480GB Mac...   
11301    BOO0089  Booq Viper Hardcase Case Macbook Retina 12 "Gr...   
777      KIN0086     Kingston DataTraveler G4 PenDrive 64GB USB 3.0   
16722    APP2253  Apple MacBook Air 13 "Core i5 18GHz | 8GB RAM ...   
18550    AP20421         Like new - Apple iPhone 8 Plus 64GB Silver   
15651    SXA0020             SecurityXtra Akimbo iPad Support Black   
13265    PUR0150  Pure Nude Ultraslim 03 "7/8 Transparent iPhone...   
17191    AP20247  Like new - Apple MacBook Pro Retina 13 "Core i...   
3150     APP1190                   IPhone AppleCare Protection Plan   
18005    MOS0234                   Moshi iGlaze iPhone Case X Black   
15784    SNN0054  Sonnet Twin 10G SFP + 10G SFP + adapter chassi...   
10956  BEL0169-A  (Open) Belkin WeMO Led Lighting LED Bulbs Sept...   
2345   ADN0016-A  Open - Adonit Jot Touch with PixelPoint pointe...   
13320    UAG0052  Urban Armor Gear Case iPhone 7 Plus Pathfinder...   
16559    APP2083  Apple MacBook Pro 13 "Core i5 with Touch Bar 3...   
16188    PAC2100  Apple iMac 27 "Core i5 3.5GHz Retina 5K | 16GB...   
398      REP0092                        Repair Speaker lower iPad 2   
2545     NAT0024  Native Union Jump battery Ligtning Cable Gray ...   
3001     SAN0098  SanDisk Connect Wireless Flash Drive 32GB Wire...   
8355     APP1322  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 8GB ...   
12851    TWS0103  "Twelve South BookBook Brown Travel Journal iP...   
16989  BEL0102-A  Open - Belkin 3 Cases iPhone Screen antihuella...   
12585    IFX0094                     iFixit Piece Speaker iPhone 5s   
14528    PAC1560      Macally External Hard Drive 500GB 25 "USB 3.0   
11761    PAC1739        QNAP TS-128 Server l Nas 8TB (1x8TB) WD Red   
18253    AP20371         Like new - Apple iPhone 7 Plus 256GB Black   
18756  NTE0007-A  Open - NewerTech AdaptaDrive adapter 2.5 "to 3...   
15530  LEX0028-A  Open - Lexar JumpDrive Lightning USB 3.0 C20i ...   
14809    GRT0457  Griffin Travel Battery Backup Power Bank for A...   

                                                    desc    price promo_price  \
17729  Robot two-wheeled personal transport up to 18K...   399.00   3.799.896   
17839  Scalable NAS server with transcoding 4K: 4-cor...  1241.96   9.621.787   
11928       Refurbished adapter mini DisplayPort to DVI.    29.99       7.607   
13579  Apple Watch 38 mm with GPS dual-core processor...   769.00  

So we were correct, over 90% of the data in this column is corrupt. There's no point deleting all of these rows, then we would barely have a products table. Instead, as it's only this column that appears to be very untrustworthy, we will delete the column.

In [64]:
products = products.drop(columns=["promo_price"])

Obviously, there's now no need to convert `promo_price` to a numerical datatype

Don't forget to download/save your new DataFrames. Also, give them an obvious name, so that you know they are the cleaned version and not the original DataFrame.

In [62]:
# path for clean data
path_clean_data = "C:/Users/ruben/WBS/Data_Science/Section_3/Data_Cleaning_Storytelling_with_Pandas_and_Matplotlib/data/clean_data/"

In [63]:
orders.to_csv(path_clean_data + "orders_cl.csv", index=False)

In [65]:
orderlines.to_csv(path_clean_data + "orderlines_cl.csv", index=False)

In [66]:
products.to_csv(path_clean_data + "products_cl.csv", index=False)